test/models/host_test.rb

```diff
@@ -1604,6 +1604,77 @@ def teardown
       assert_equal 2, host.interfaces.size
     end
 
+    context 'update_subnets_from_facts is enabled' do
+      setup do
+        Setting[:update_subnets_from_facts] = true
+      end
+
+      test "#set_interfaces updates existing physical interface when subnet does not match ip and new subnet is unknown" do
+        subnet = FactoryGirl.create(:subnet_ipv4, :dhcp, :network => '10.10.0.0')
+        host = FactoryGirl.create(:host,
+                                  :managed,
+                                  :mac => '00:00:00:11:22:33',
+                                  :ip => '10.10.0.1',
+                                  :subnet => subnet
+                                 )
+        hash = { :eth0 => {:macaddress => '00:00:00:11:22:33', :ipaddress => '10.10.20.2', :virtual => false}
+        }.with_indifferent_access
+        parser = stub(:interfaces => hash, :ipmi_interface => {}, :suggested_primary_interface => hash.to_a.last)
+
+        assert_no_difference 'Nic::Base.count' do
+          host.set_interfaces(parser)
+        end
+
+        assert_equal '10.10.20.2', host.primary_interface.ip
+        assert_nil host.primary_interface.subnet
+      end
+
+      test "#set_interfaces updates existing physical interface when subnet does not match ip and new subnet is known" do
+        subnet = FactoryGirl.create(:subnet_ipv4, :dhcp, :network => '10.10.0.0')
+        subnet2 = FactoryGirl.create(:subnet_ipv4, :dhcp, :network => '10.10.20.0')
+        host = FactoryGirl.create(:host,
+                                  :managed,
+                                  :mac => '00:00:00:11:22:33',
+                                  :ip => '10.10.0.1',
+                                  :subnet => subnet
+                                 )
+        hash = { :eth0 => {:macaddress => '00:00:00:11:22:33', :ipaddress => '10.10.20.2', :virtual => false}
+        }.with_indifferent_access
+        parser = stub(:interfaces => hash, :ipmi_interface => {}, :suggested_primary_interface => hash.to_a.last)
+
+        assert_no_difference 'Nic::Base.count' do
+          host.set_interfaces(parser)
+        end
+
+        assert_equal '10.10.20.2', host.primary_interface.ip
+        assert_equal subnet2, host.primary_interface.subnet
+      end
+
+      test "#set_interfaces updates existing physical interface when subnet6 is set but facts report no ipv6 addr" do
+        subnet = FactoryGirl.create(:subnet_ipv4, :network => '10.10.10.0')
```

Comment:
Looks like the `subnet` field is the same in all tests, please move it to the setup method